File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
       ListNode reverse = null;
       
       ListNode curr = head;
       
       while(curr != null) {
        
        ListNode second = curr.next;
        curr.next = reverse;
        reverse = curr;
        curr = second;

       }
 
       return reverse; 
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # optimal runtime O(n + m) 
        # we're iterating through both lists
        list3 = ListNode()
        # dummy node to modify list3, in-place
        temp = list3 
        while list1 and list2:
            if list1.val > list2.val:
                temp.next = list2
                list2 = list2.next
            else:
                temp.next = list1
                list1 = list1.next
            temp = temp.next
        
        while list1:
            temp.next = list1
            list1 = list1.next
            temp = temp.next

        while list2:
            temp.next = list2
            list2 = list2.next
            temp = temp.next
        
        return list3.next




        

        